crates/services/txpool_v2/src/storage/graph.rs

```diff
@@ -0,0 +1,477 @@
+                .extend(self.remove_node_and_dependent_sub_graph(collision)?);
+        }
+        // Add the new transaction to the graph and update the others in consequence
+        let tip = transaction.tip();
+        let gas = transaction.max_gas();
+        let outputs = transaction.outputs().clone();
+        let node = StorageData {
+            dependents_cumulative_tip: tip,
+            dependents_cumulative_gas: gas,
+            transaction,
+            number_txs_in_chain: 1,
+            submitted_time: Instant::now(),
+        };
+
+        let mut whole_tx_chain = vec![];
+
+        // Check if the dependency chain is too big
+        let mut to_check = dependencies.clone();
+        while let Some(node_id) = to_check.pop() {
+            let Some(dependency_node) = self.graph.node_weight(node_id) else {
+                return Err(Error::Storage(format!(
+                    "Node with id {:?} not found",
+                    node_id
+                )));
+            };
+            if dependency_node.number_txs_in_chain >= self.config.max_dependent_txn_count
+            {
+                return Err(Error::NotInsertedChainDependencyTooBig);
+            }
+            whole_tx_chain.push(node_id);
+            to_check.extend(self.get_dependencies(node_id)?);
+        }
+
+        // Add the transaction to the graph
+        let node_id = self.graph.add_node(node);
+        for dependency in dependencies {
+            self.graph.add_edge(dependency, node_id, ());
+        }
+        self.cache_tx_infos(&outputs, &tx_id, node_id)?;
+
+        // Update the cumulative tip and gas of the dependencies transactions and recursively their dependencies, etc.
+        for node_id in whole_tx_chain {
+            let Some(node) = self.graph.node_weight_mut(node_id) else {
+                return Err(Error::Storage(format!(
+                    "Node with id {:?} not found",
+                    node_id
+                )));
+            };
+            node.number_txs_in_chain = node.number_txs_in_chain.saturating_add(1);
+            node.dependents_cumulative_tip =
+                node.dependents_cumulative_tip.saturating_add(tip);
+            node.dependents_cumulative_gas =
+                node.dependents_cumulative_gas.saturating_add(gas);
+        }
+        Ok((node_id, removed_transactions))
+    }
+
+    fn get(&self, index: &Self::StorageIndex) -> Result<&StorageData, Error> {
+        self.get_inner(index)
+    }
+
+    fn get_dependencies(
+        &self,
+        index: Self::StorageIndex,
+    ) -> Result<Vec<Self::StorageIndex>, Error> {
+        Ok(self
+            .graph
+            .neighbors_directed(index, petgraph::Direction::Incoming)
+            .collect())
+    }
+
+    fn get_dependents(
+        &self,
+        index: Self::StorageIndex,
+    ) -> Result<Vec<Self::StorageIndex>, Error> {
+        self.get_dependents_inner(&index)
+    }
+
+    fn collect_dependencies_transactions(
```

Comment:
I agree with your point. However, I placed them together because in both cases we need to iterate through the inputs and I wanted to avoid doing it twice.
Given the context, do you think it's non-comprehensive and premature optimization and I should still do your proposition ? (In this case I'll do it)

Reply:
I think it is better to split this functionality. Yeah, we need to iterate over inputs twice, but only in one case we will touch database.

Plus this function can be used later in https://github.com/FuelLabs/fuel-core/pull/2193

Reply:
Done